### PR TITLE
[3006.x] fix(saltclass): don't lose nested classes and states in recursion

### LINKED ISSUE #58969: [BUG] saltclass tops state list is incomplete for nested classes
**Description**
When using saltclass with multiple levels of classes, all of which define states, the resulting state list only comes back with states for the top level classes.  Looking at the code, it appears that it is ignoring the state lists returned from the recursive calls of `expand_classes_in_order`

**Setup**
```
Node: 
  classes: [A,L]

Class A:
  states: [1,2]
  classes: [B]
Class B:
  states: [3,4]
  classes: [C]
Class C:
  states: [5,6]
 
Class: L
  states: [7]
  classes: [M]
Class M:
  states: [8]
```

**Expected behavior**
Running state.show_top with the above classes, one would expect a state list of `[1,2,3,4,5,6,7,8]`, but what comes back instead is `[1,2,7]`

**Versions Report**
<details><summary>salt --versions-report</summary>

```
Salt Version:
           Salt: 3002.1
 
Dependency Versions:
           cffi: 1.14.2
       cherrypy: Not Installed
       dateutil: 2.8.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.11.1
        libgit2: 0.28.2
       M2Crypto: 0.35.2
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: 2.20
       pycrypto: 2.6.1
   pycryptodome: 3.9.7
         pygit2: 0.28.2
         Python: 3.6.8 (default, Oct 13 2020, 16:18:22)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.0.0
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-1160.2.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7 Core

```
</details>

**Additional context**
Without knowing exactly what was intended, this is my rough guess as to what might have been meant to happen:
```
--- salt/utils/saltclass.py        2020-11-18 04:04:34.890045089 +0000
+++ -   2020-11-18 04:04:47.691585445 +0000
@@ -311,7 +311,7 @@
             if expanded_classes[klass].get("classes"):
                 l_id = classes_to_expand.index(klass)
                 classes_to_expand[l_id:l_id] = expanded_classes[klass]["classes"]
-                expand_classes_in_order(
+                _,_,minion_dict["states"] = expand_classes_in_order(
                     minion_dict,
                     salt_data,
                     seen_classes,
@@ -319,7 +319,7 @@
                     classes_to_expand,
                 )
             else:
-                expand_classes_in_order(
+                _,_,minion_dict["states"] = expand_classes_in_order(
                     minion_dict,
                     salt_data,
                     seen_classes,
@@ -347,7 +347,7 @@
             "states" in expanded_classes[ord_klass]
             and expanded_classes[ord_klass]["states"] is None
         ):
-            expanded_classes[ord_klass]["states"] = {}
+            expanded_classes[ord_klass]["states"] = []
 
         if "states" in expanded_classes[ord_klass]:
             ord_expanded_states.extend(expanded_classes[ord_klass]["states"])
@@ -357,7 +357,9 @@
         minion_dict["states"] = []
 
     if "states" in minion_dict:
-        ord_expanded_states.extend(minion_dict["states"])
+        for state in minion_dict["states"]:
+            if state not in ord_expanded_states:
+                ord_expanded_states.append(state)
 
     ord_expanded_classes.append(minion_dict)

```

### LINKED ISSUE #63933: [BUG] saltclass node pillar does not override classes pillar dict
**Description**

I started using saltclass for tops and pillars and I am not sure, how the merging of pillar data works (especially dict merging).

**Setup**
I used a minion test installation (KVM, Debian 11) and a FreeBSD machine for master with a test class and a test node.

**Steps to Reproduce the behavior**

Class definition (`saltclass1`):

```
pillars:
  dict1:
    element: value from class
```

Node definition:

```
classses:
  - saltclass1
pillars:
  dict1:
    element: value from node
```

**Expected behavior**

After updating pillars with `salt 'node' saltutil.refresh_pillar` I expected that node pillar data overrides classes pillar data, like:

```bash
salt 'saltclass*' pillar.items

node:
    ----------
    __saltclass__:
        ----------
        classes:
            - saltclass1
        environment:
            base
        nodename:
            node
        states:
    dict1:
        ----------
        element:
            value from node
```

What I got instead is:

```bash
salt 'saltclass*' pillar.items

node:
    ----------
    __saltclass__:
        ----------
        classes:
            - saltclass1
        environment:
            base
        nodename:
            node
        states:
    dict1:
        ----------
        element:
            value from class    <--- not overriden from node value
```

**Versions Report**
For master: 

```yaml
Salt Version:
          Salt: 3005
 
Dependency Versions:
          cffi: 1.15.1
      cherrypy: Not Installed
      dateutil: Not Installed
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.0.1
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.4
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.15.0
        pygit2: Not Installed
        Python: 3.9.15 (main, Nov 17 2022, 01:10:49)
  python-gnupg: Not Installed
        PyYAML: 5.4.1
         PyZMQ: 22.3.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.4
 
System Versions:
          dist: freebsd 13.1 
        locale: utf-8
       machine: amd64
       release: 13.1-RELEASE-p3
        system: FreeBSD
       version: FreeBSD 13.1 
```

Minion is 3002 (Debian package).

**Additional context**

The original reclass comment that is not very precise how the merging is done: "After all classes (and the classes they reference) have been visited, **reclass** finally merges the applications list and parameters defined for the node into what has been accumulated during the processing of the classes, and returns the final result."

Thanks
Meik

### PR DESCRIPTION
**Pull request by baby-gnu**
### What does this PR do?

This PR keep all the classes and states from recursive expansion of saltclass classes.

### What issues does this PR fix or reference?
Fixes #58969
Fixes #63933

### Previous Behavior

Only states from top level classes were kept.

### New Behavior

All states from all classes are used, in the correct order.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices, including the
[PR Guidelines](https://docs.saltproject.io/en/master/topics/development/pull_requests.html).

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: tools/precommit/docstrings.py [hunk 1]
```diff
@@ -345,7 +345,6 @@
         "get_pillars",
         "expand_variables",
         "render_jinja",
-        "expand_classes_in_order",
```
Comment (twangboy) [inline tools/precommit/docstrings.py:348]: Why is this removed?

**Comment by twangboy** [comment]
We'll need a changelog and some tests for this.

File: tools/precommit/docstrings.py (hunk 1 above)
Comment (baby-gnu) [reply tools/precommit/docstrings.py:348]: Because I added a docsting to `expand_classes_in_order`.

When I looked up this function name in a global grep to see if any API need to be changed I found it there and I thought it was a good thing to do but I can remove that commit if it's inappropriate.

**Comment by baby-gnu** [comment]
> We'll need a changelog and some tests for this.

Ok, the changelog is quite easy, I'll need to document for testing.

**Review by twangboy — DISMISSED** [review]

File: tools/precommit/docstrings.py (hunk 1 above)
Comment (twangboy) [reply tools/precommit/docstrings.py:348]: Oh, my bad... this is right

**Comment by baby-gnu** [comment]
> We'll need a changelog and some tests for this.

@twangboy: I added tests for both issues:

<details><summary>Errors when passing the tests before my PR<code>nox -e 'test-3(coverage=False)' -- tests/pytests/unit/pillar/test_saltclass.py</code></summary>

```
nox > Running session test-3(coverage=False)
nox > Re-using existing virtual environment at .nox/test-3-coverage-false.
nox > Session test-3(coverage=False) was successful.
nox > Running session test-parametrized-3(crypto=None, transport='zeromq', coverage=False)
nox > Re-using existing virtual environment at .nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false.
nox > python -m pip install --progress-bar=off -U setuptools pip wheel
nox > python -m pip install --progress-bar=off -r requirements/static/ci/py3.10/linux.txt
nox > python -m pytest --rootdir /src --log-file-level=debug --show-capture=no -ra -s -vv --showlocals --log-file=/src/artifacts/logs/runtests-20240927064217.175879.log --transport=zeromq tests/pytests/unit/pillar/test_saltclass.py
<frozen importlib._bootstrap>:914: ImportWarning: TornadoImporter.find_spec() not found; falling back to find_module()
<frozen importlib._bootstrap>:914: ImportWarning: NaclImporter.find_spec() not found; falling back to find_module()
============================= test session starts ==============================
platform linux -- Python 3.10.14, pytest-8.1.1, pluggy-1.4.0 -- /src/.nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/bin/python
cachedir: .pytest_cache
max open files: soft=1048576; hard=1048576
salt-transport: zeromq
rootdir: /src
configfile: pytest.ini
plugins: skip-markers-1.5.2, httpserver-1.0.8, helpers-namespace-2021.4.29, flaky-3.8.1, salt-factories-1.0.1, timeout-2.3.1, pyfakefs-5.3.1, system-statistics-1.0.2, custom-exit-code-0.3.0, subtests-0.4.0, shell-utilities-1.8.0
collecting ... collected 7 items

tests/pytests/unit/pillar/test_saltclass.py::test_classes_order FAILED
tests/pytests/unit/pillar/test_saltclass.py::test_list_expansion_succeeds PASSED
tests/pytests/unit/pillar/test_saltclass.py::test_node_override_classes_scalars PASSED
tests/pytests/unit/pillar/test_saltclass.py::test_node_override_classes_scalar_in_dict FAILED
tests/pytests/unit/pillar/test_saltclass.py::test_node_override_classes_list_in_dict PASSED
tests/pytests/unit/pillar/test_saltclass.py::test_list_in_dict_no_duplication FAILED
tests/pytests/unit/pillar/test_saltclass.py::test_nested_classes_has_pillars PASSED

=================================== FAILURES ===================================
______________________________ test_classes_order ______________________________

temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_classes_order0/saltclass/examples')

    def test_classes_order(temp_saltclass_tree):
        """
        Classes must be correctly ordered.
    
        See: https://github.com/saltstack/salt/issues/58969
        """
        expected_ret = [
            "default.users.foo",
            "default.users",
            "default.motd",
            "default.empty",
            "default",
            "roles.app",
            "roles.nginx",
        ]
        fake_args = {"path": str(temp_saltclass_tree)}
        fake_pillar = {}
        fake_minion_id = "fake_id"
        try:
            full_ret = saltclass.ext_pillar(fake_minion_id, fake_pillar, fake_args)
            parsed_ret = full_ret["__saltclass__"]["classes"]
        # Fail the test if we hit our NoneType error
        except TypeError as err:
            pytest.fail(err)
        # Else give the parsed content result
>       assert expected_ret == parsed_ret
E       AssertionError: assert ['default.users.foo', 'default.users', 'default.motd', 'default.empty', 'default', 'roles.app', 'roles.nginx'] == ['default.users', 'default.motd', 'default.empty', 'default', 'roles.app', 'roles.nginx']
E         
E         At index 0 diff: 'default.users.foo' != 'default.users'
E         Left contains one more item: 'roles.nginx'
E         
E         Full diff:
E           [
E         +     'default.users.foo',
E               'default.users',
E               'default.motd',
E               'default.empty',
E               'default',
E               'roles.app',
E               'roles.nginx',
E           ]

expected_ret = ['default.users.foo',
 'default.users',
 'default.motd',
 'default.empty',
 'default',
 'roles.app',
 'roles.nginx']
fake_args  = {'path': '/tmp/pytest-of-root/pytest-77/test_classes_order0/saltclass/examples'}
fake_minion_id = 'fake_id'
fake_pillar = {'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from nginx',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
full_ret   = {'__saltclass__': {'classes': ['default.users',
                               'default.motd',
                               'default.empty',
                               'default',
                               'roles.app',
                               'roles.nginx'],
                   'environment': 'base',
                   'nodename': 'fake_id',
                   'states': ['users',
                              'motd',
                              'default',
                              'app',
                              'nginx',
                              'minion_node']},
 'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from minion_node',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
parsed_ret = ['default.users',
 'default.motd',
 'default.empty',
 'default',
 'roles.app',
 'roles.nginx']
temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_classes_order0/saltclass/examples')

tests/pytests/unit/pillar/test_saltclass.py:174: AssertionError
__________________ test_node_override_classes_scalar_in_dict ___________________

temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_node_override_classes_sca1/saltclass/examples')

    def test_node_override_classes_scalar_in_dict(temp_saltclass_tree):
        """
        Scalars defined in `dict` pillars defined in a node definition must override the
        same dict definition from classes.
    
        See: https://github.com/saltstack/salt/issues/63933
        """
        expected_ret = "from minion_node"
        fake_args = {"path": str(temp_saltclass_tree)}
        fake_pillar = {}
        fake_minion_id = "fake_id"
        try:
            full_ret = saltclass.ext_pillar(fake_minion_id, fake_pillar, fake_args)
            parsed_ret = full_ret["test_dict"]["a_scalar"]
        # Fail the test if we hit our NoneType error
        except TypeError as err:
            pytest.fail(err)
        # Else give the parsed content result
>       assert expected_ret == parsed_ret
E       AssertionError: assert 'from minion_node' == 'from users'
E         
E         - from users
E         + from minion_node

expected_ret = 'from minion_node'
fake_args  = {'path': '/tmp/pytest-of-root/pytest-77/test_node_override_classes_sca1/saltclass/examples'}
fake_minion_id = 'fake_id'
fake_pillar = {'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from nginx',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
full_ret   = {'__saltclass__': {'classes': ['default.users',
                               'default.motd',
                               'default.empty',
                               'default',
                               'roles.app',
                               'roles.nginx'],
                   'environment': 'base',
                   'nodename': 'fake_id',
                   'states': ['users',
                              'motd',
                              'default',
                              'app',
                              'nginx',
                              'minion_node']},
 'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from minion_node',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
parsed_ret = 'from users'
temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_node_override_classes_sca1/saltclass/examples')

tests/pytests/unit/pillar/test_saltclass.py:231: AssertionError
_______________________ test_list_in_dict_no_duplication _______________________

temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_list_in_dict_no_duplicati0/saltclass/examples')

    def test_list_in_dict_no_duplication(temp_saltclass_tree):
        """
        `list` under a `dict` in pillar must not be duplicated.
    
        See:
        """
        expected_ret = ["element1", "element2"]
        fake_args = {"path": str(temp_saltclass_tree)}
        fake_pillar = {}
        fake_minion_id = "fake_id"
        try:
            full_ret = saltclass.ext_pillar(fake_minion_id, fake_pillar, fake_args)
            parsed_ret = full_ret["test_dict"]["a_list"]
        # Fail the test if we hit our NoneType error
        except TypeError as err:
            pytest.fail(err)
        # Else give the parsed content result
>       assert expected_ret == parsed_ret
E       AssertionError: assert ['element1', 'element2'] == ['element1', 'element2', 'element1', 'element2']
E         
E         Right contains 2 more items, first extra item: 'element1'
E         
E         Full diff:
E           [
E               'element1',
E               'element2',
E         -     'element1',
E         -     'element2',
E           ]

expected_ret = ['element1', 'element2']
fake_args  = {'path': '/tmp/pytest-of-root/pytest-77/test_list_in_dict_no_duplicati0/saltclass/examples'}
fake_minion_id = 'fake_id'
fake_pillar = {'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from nginx',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
full_ret   = {'__saltclass__': {'classes': ['default.users',
                               'default.motd',
                               'default.empty',
                               'default',
                               'roles.app',
                               'roles.nginx'],
                   'environment': 'base',
                   'nodename': 'fake_id',
                   'states': ['users',
                              'motd',
                              'default',
                              'app',
                              'nginx',
                              'minion_node']},
 'default': {'network': {'dns': None,
                         'ntp': {'srv1': '192.168.10.10',
                                 'srv2': '192.168.10.20'}},
             'ntp': {'srv1': '192.168.20.10'}},
 'global_scalar': 'from minion_node',
 'nginx_scalar': 'from nginx',
 'test_dict': {'a_list': ['element1', 'element2', 'element1', 'element2'],
               'a_scalar': 'from users'},
 'test_list': [{'a': '192.168.10.10'}, '192.168.10.20']}
parsed_ret = ['element1', 'element2', 'element1', 'element2']
temp_saltclass_tree = PosixPath('/tmp/pytest-of-root/pytest-77/test_list_in_dict_no_duplicati0/saltclass/examples')

tests/pytests/unit/pillar/test_saltclass.py:272: AssertionError
=========================== short test summary info ============================
FAILED tests/pytests/unit/pillar/test_saltclass.py::test_classes_order - Asse...
FAILED tests/pytests/unit/pillar/test_saltclass.py::test_node_override_classes_scalar_in_dict
FAILED tests/pytests/unit/pillar/test_saltclass.py::test_list_in_dict_no_duplication
========================= 3 failed, 4 passed in 1.09s ==========================
nox > Command python -m pytest --rootdir /src --log-file-level=debug --show-capture=no -ra -s -vv --showlocals --log-file=/src/artifacts/logs/runtests-20240927064217.175879.log --transport=zeromq tests/pytests/unit/pillar/test_saltclass.py failed with exit code 1
nox > Session test-parametrized-3(crypto=None, transport='zeromq', coverage=False) failed.
nox > Ran multiple sessions:
nox > * test-3(coverage=False): success
nox > * test-parametrized-3(crypto=None, transport='zeromq', coverage=False): failed
```
</details>

**Review by twangboy — DISMISSED** [review]

**Comment by baby-gnu** [comment]
Thanks @twangboy, should I rebase the branch over 3006.x from time to time until it get merge?

**Comment by baby-gnu** [comment]
> Thanks @twangboy, should I rebase the branch over 3006.x from time to time until it get merge?

I reword my commit message to fix typo as you did in the title :+1:

**Review by twangboy — APPROVED** [review]
